packages/circuits/jwt-verifier-template.circom

```diff
@@ -0,0 +1,326 @@
+    iss <== Bytes2Ints(issLen)(issMatch);
+
+    // Verify if the key `iat` in the payload is unique
+    var iatKeyLength = IAT_KEY_LENGTH();
+    var iatKey[iatKeyLength] = IAT_KEY();
+    signal iatKeyMatch[iatKeyLength] <== RevealSubstring(maxPayloadLength, iatKeyLength, 1)(payload, iatKeyStartIndex, iatKeyLength);
+    for (var i = 0; i < iatKeyLength; i++) {
+        iatKeyMatch[i] === iatKey[i];
+    }
+
+    // Reveal the iat in the payload
+    var iatLength = TIMESTAMP_LENGTH();
+    signal iatStartIndex <== iatKeyStartIndex + iatKeyLength;
+    signal iatMatch[iatLength] <== RevealSubstring(maxPayloadLength, iatLength, 0)(payload, iatStartIndex, iatLength);
+    timestamp <== Digit2Int(iatLength)(iatMatch);
+
+    // Verify if the key `azp` in the payload is unique
+    var azpKeyLength = AZP_KEY_LENGTH();
+    var azpKey[azpKeyLength] = AZP_KEY();
+    signal azpKeyMatch[azpKeyLength] <== RevealSubstring(maxPayloadLength, azpKeyLength, 1)(payload, azpKeyStartIndex, azpKeyLength);
+    for (var i = 0; i < azpKeyLength; i++) {
+        azpKeyMatch[i] === azpKey[i];
+    }
+
+    // Reveal the azp in the payload
+    signal azpStartIndex <== azpKeyStartIndex + azpKeyLength + 1;
+    signal azpMatch[maxAzpLength] <== RevealSubstring(maxPayloadLength, maxAzpLength, 0)(payload, azpStartIndex, azpLength);
+    azp <== Bytes2Ints(maxAzpLength)(azpMatch);
+
+    // Verify if the key `email` in the payload is unique
+    var emailKeyLength = EMAIL_KEY_LENGTH();
+    var emailKey[emailKeyLength] = EMAIL_KEY();
+    signal emailKeyMatch[emailKeyLength] <== RevealSubstring(maxPayloadLength, emailKeyLength, 1)(payload, emailKeyStartIndex, emailKeyLength);
+    for (var i = 0; i < emailKeyLength; i++) {
+        emailKeyMatch[i] === emailKey[i];
+    }
+
+    // Reveal the email in the payload
+    var maxEmailLength = EMAIL_ADDR_MAX_BYTES();
+    signal emailStartIndex <== emailKeyStartIndex + emailKeyLength + 1;
+    signal email[maxEmailLength] <== RevealSubstring(maxPayloadLength, maxEmailLength, 0)(payload, emailStartIndex, emailLength);
+
+    // Calculate account salt using email
+    var numEmailInts = compute_ints_size(maxEmailLength);
+    signal emailInts[numEmailInts] <== Bytes2Ints(maxEmailLength)(email);
+    accountSalt <== AccountSalt(numEmailInts)(emailInts, accountCode);
+
+    // Verify if the key `nonce` in the payload is unique
+    var nonceKeyLength = NONCE_LENGTH();
+    var nonceKey[nonceKeyLength] = NONCE();
+    signal nonceKeyMatch[nonceKeyLength] <== RevealSubstring(maxPayloadLength, nonceKeyLength, 1)(payload, nonceKeyStartIndex, nonceKeyLength);
+    for (var i = 0; i < nonceKeyLength; i++) {
+        nonceKeyMatch[i] === nonceKey[i];
+    }
+
+    // Reveal the command in the nonce
+    signal commandStartIndex <== nonceKeyStartIndex + nonceKeyLength + 1;
+    signal command[maxCommandLength] <== RevealSubstring(maxPayloadLength, maxCommandLength, 0)(payload, commandStartIndex, commandLength);
```

Comment:
I don't understand how this line makes a constraint that the command must exist between "<div id=3D"zkemail">" and "</div>".

Reply:
There is no constraint that the actual command is extracted as it is not possible to this check when we are extracting substrings. But this is secure because even if the `commandLength` is wrong, the verification step will fail as it will give an invalid command.

Reply:
Then how does the circuit prevent JWT issued for applications not for zkemail from being used for proof generation?
In other words, an adversary can use any (existing) JWT issued by google, if the circuit doesn't require any zkemail-specific string in the command part.

Reply:
There's a field named `azp` Authorised Party which mentions the app the JWT was signed for, which will help us identify that the JWT is not malicious.